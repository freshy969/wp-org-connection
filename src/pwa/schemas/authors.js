import { schema } from 'normalizr';

export const author = new schema.Entity(
  'author',
  {},
  {
    processStrategy(entity) {
      entity.mst = 'author';
      entity.type = 'author';
      return entity;
    },
  },
);
export const authors = new schema.Array(author);
