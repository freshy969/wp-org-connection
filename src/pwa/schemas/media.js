import { schema } from 'normalizr';

export const media = new schema.Entity('media');
export const mediaArray = new schema.Array(media);
