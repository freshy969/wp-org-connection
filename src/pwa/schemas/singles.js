import { schema } from 'normalizr';

const author = new schema.Entity('author');
const media = new schema.Entity('media');
const taxonomy = new schema.Entity('taxonomy');
const taxonomies = new schema.Array(new schema.Array(taxonomy));
export const single = new schema.Entity('single')
single.define({
  _embedded: {
    author: [author],
    'wp:featuredmedia': [media],
    'wp:term': taxonomies,
    up: single,
  },
})
export const singles = new schema.Array(single);
