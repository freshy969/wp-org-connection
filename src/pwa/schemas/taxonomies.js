import { schema } from 'normalizr';

export const taxonomy = new schema.Entity('taxonomy');
export const taxonomies = new schema.Array(taxonomy);
