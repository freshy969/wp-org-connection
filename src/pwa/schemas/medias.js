import { schema } from 'normalizr';
import { author } from './authors';

export const media = new schema.Entity(
  'media',
  {
    _embedded: {
      author: [author],
    },
  },
  {
    processStrategy(entity) {
      entity.mst = 'media';
      return entity;
    },
  },
);

export const medias = new schema.Array(media);
