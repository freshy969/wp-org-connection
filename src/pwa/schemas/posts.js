/* eslint-disable no-underscore-dangle */
import { schema } from 'normalizr';
import { taxonomy } from './taxonomies';
import { author } from './authors';

const media = new schema.Entity('media');
const taxonomies = new schema.Array(new schema.Array(taxonomy));
export const post = new schema.Entity(
  'post',
  {},
  {
    processStrategy(entity) {
      // Get all taxonomies and generate a map so we can now later on which props are actually
      // taxonomies.
      if (entity._embedded && entity._embedded['wp:term']) {
        const taxonomiesMap = {};
        entity._embedded['wp:term'].forEach(term =>
          term.forEach(item => {
            const type = item.taxonomy === 'post_tag' ? 'tag' : item.taxonomy;
            taxonomiesMap[type] = taxonomiesMap[type] || [];
            taxonomiesMap[type].push(item.id);
          }),
        );
        return { ...entity, taxonomiesMap };
      }
      return entity;
    },
  },
);
post.define({
  _embedded: {
    author: [author],
    'wp:featuredmedia': [media],
    'wp:term': taxonomies,
    up: post,
  },
});
export const posts = new schema.Array(post);
