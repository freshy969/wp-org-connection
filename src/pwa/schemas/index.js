import { post, posts } from './posts';
import { taxonomy, taxonomies } from './taxonomies';
import { author, authors } from './authors';

export {
  post,
  posts,
  taxonomy,
  taxonomies,
  author,
  authors,
};
