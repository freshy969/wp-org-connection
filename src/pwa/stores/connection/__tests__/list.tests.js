/* eslint-disable dot-notation */
import { autorun } from 'mobx';
import { types, unprotect } from 'mobx-state-tree';
import { normalize } from 'normalizr';
import * as connect from '../';
import { list } from '../../../schemas';
import postsFromCategory7 from '../../../__tests__/posts-from-category-7.json';
import postsFromCategory7Page2 from '../../../__tests__/posts-from-category-7-page-2.json';

const { result: resultFromCategory7, entities: entitiesFromCategory } = normalize(
  postsFromCategory7,
  list,
);

const { result: resultFromCategory7Page2, entities: entitiesFromCategoryPage2 } = normalize(
  postsFromCategory7Page2,
  list,
);

const Connection = types
  .model()
  .props(connect.props)
  .views(connect.views)
  .actions(connect.actions);

let connection = null;
beforeEach(() => {
  connection = Connection.create({});
  unprotect(connection);
});

describe('Store â€º List', () => {
  test('Get list snapshot when entity is not ready', () => {
    expect(connection.list('category', 7)).toMatchSnapshot();
  });

  test('Get list shape when entity is not ready', () => {
    expect(connection.list('category', 7).ready).toBe(false);
    expect(connection.list('category', 7).pages).toEqual([]);
    expect(connection.list('category', 7).page(2).ready).toBe(false);
    expect(connection.list('category', 7).page(2).entities).toEqual([]);
    expect(connection.list('category', 7).pages).toEqual([]);
    expect(connection.list('category', 7).entity.id).toBe(7);
    expect(connection.list('category', 7).entity.link).toBe('/?cat=7');
    expect(connection.list('tag', 10).entity.link).toBe('/?tag_ID=10');
    expect(connection.list('latest', 'post').entity.link).toBe('/');
    expect(connection.list('category', 7).entity.pagedLink(7)).toBe('/?cat=7&paged=7');
    expect(connection.list('latest', 'post').entity.pagedLink(7)).toBe('/page/7');
    expect(connection.list('author', 4).entity.avatar).toBe('');
  });

  test('Get list entity shapes after adding a page', () => {
    expect(connection.list('category', 7).entities).toEqual([]);
    expect(connection.list('category', 7).page(1).entities).toEqual([]);
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
    });
    const results = [57, 54, 42, 36, 33];
    expect(connection.list('category', 7).entities.map(entity => entity.id)).toEqual(results);
    expect(
      connection
        .list('category', 7)
        .page(1)
        .entities.map(entity => entity.id),
    ).toEqual(results);
    expect(connection.list('category', 7).entities[0].id).toBe(57);
    expect(connection.list('category', 7).entities[0].title).toBe('Shinjuku Gyoen National Garden');
    expect(connection.list('category', 7).page(1).entities[0].id).toBe(57);
    expect(connection.list('category', 7).page(1).entities[0].title).toBe(
      'Shinjuku Gyoen National Garden',
    );
    expect(connection.list('category', 7).pages[0].entities[0].id).toBe(57);
    expect(connection.list('category', 7).pages[0].entities[0].title).toBe(
      'Shinjuku Gyoen National Garden',
    );
  });

  test('Get list entity shapes after adding two pages', () => {
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
    });
    connection.addPage({
      type: 'category',
      id: 7,
      page: 2,
      result: resultFromCategory7Page2,
      entities: entitiesFromCategoryPage2,
    });
    const page1 = [57, 54, 42, 36, 33];
    const page2 = [30, 26, 23, 12, 1];
    expect(connection.list('category', 7).entities.map(entity => entity.id)).toEqual([
      ...page1,
      ...page2,
    ]);
    expect(
      connection
        .list('category', 7)
        .page(2)
        .entities.map(entity => entity.id),
    ).toEqual(page2);
    expect(connection.list('category', 7).entities[5].id).toBe(30);
    expect(connection.list('category', 7).entities[5].title).toBe('Machu Picchu');
    expect(connection.list('category', 7).page(2).entities[0].id).toBe(30);
    expect(connection.list('category', 7).page(2).entities[0].title).toBe('Machu Picchu');
    expect(connection.list('category', 7).pages[1].entities[0].id).toBe(30);
    expect(connection.list('category', 7).pages[1].entities[0].title).toBe('Machu Picchu');
  });

  test('Subscribe to ready (in page) before entity is ready', done => {
    expect(connection.list('category', 7).page(2).ready).toBe(false);
    autorun(() => {
      if (connection.list('category', 7).page(2).ready) done();
    });
    connection.addPage({
      type: 'category',
      id: 7,
      page: 2,
      result: resultFromCategory7Page2,
      entities: entitiesFromCategoryPage2,
    });
  });

  test('Subscribe to ready before entity is ready', done => {
    expect(connection.list('category', 7).ready).toBe(false);
    autorun(() => {
      if (connection.list('category', 7).ready) done();
    });
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
    });
  });

  test('Subscribe to fetching (in page) before entity is ready', done => {
    expect(connection.list('category', 7).page(1).fetching).toBe(false);
    autorun(() => {
      if (connection.list('category', 7).page(1).fetching) done();
    });
    connection.fetchingPage({ type: 'category', id: 7, page: 1 });
    expect(connection.list('category', 7).page(1).ready).toBe(false);
  });

  test('Subscribe to fetching before entity is ready', done => {
    expect(connection.list('category', 7).fetching).toBe(false);
    autorun(() => {
      if (connection.list('category', 7).fetching) done();
    });
    connection.fetchingPage({ type: 'category', id: 7, page: 1 });
    expect(connection.list('category', 7).ready).toBe(false);
  });

  test('Ready should remain true even if new pages are fetched', () => {
    expect(connection.list('category', 7).ready).toBe(false);
    expect(connection.list('category', 7).page(1).ready).toBe(false);
    expect(connection.list('category', 7).page(2).ready).toBe(false);
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
    });
    expect(connection.list('category', 7).ready).toBe(true);
    expect(connection.list('category', 7).page(1).ready).toBe(true);
    expect(connection.list('category', 7).page(2).ready).toBe(false);
    connection.fetchingPage({ type: 'category', id: 7, page: 2 });
    expect(connection.list('category', 7).ready).toBe(true);
    expect(connection.list('category', 7).page(1).ready).toBe(true);
    expect(connection.list('category', 7).page(2).ready).toBe(false);
  });

  test('Fetching should go back to false when new pages are fetched', () => {
    expect(connection.list('category', 7).fetching).toBe(false);
    expect(connection.list('category', 7).page(1).fetching).toBe(false);
    expect(connection.list('category', 7).page(2).fetching).toBe(false);
    connection.fetchingPage({ type: 'category', id: 7, page: 1 });
    expect(connection.list('category', 7).fetching).toBe(true);
    expect(connection.list('category', 7).page(1).fetching).toBe(true);
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
    });
    expect(connection.list('category', 7).fetching).toBe(false);
    expect(connection.list('category', 7).page(1).fetching).toBe(false);
    connection.fetchingPage({ type: 'category', id: 7, page: 2 });
    expect(connection.list('category', 7).fetching).toBe(true);
    expect(connection.list('category', 7).page(1).fetching).toBe(false);
    expect(connection.list('category', 7).page(2).fetching).toBe(true);
  });

  test('Total shapes before and after initialization', () => {
    expect(connection.list('category', 7).total.entities).toBe(null);
    expect(connection.list('category', 7).total.pages).toBe(null);
    expect(connection.list('category', 7).total.fetched.entities).toBe(null);
    expect(connection.list('category', 7).total.fetched.pages).toBe(null);
    expect(connection.list('category', 7).page(1).total).toBe(null);
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
      total: { entities: 10, pages: 2 },
    });
    expect(connection.list('category', 7).total.entities).toBe(10);
    expect(connection.list('category', 7).total.pages).toBe(2);
    expect(connection.list('category', 7).page(1).total).toBe(5);
    expect(connection.list('category', 7).total.fetched.entities).toBe(5);
    expect(connection.list('category', 7).total.fetched.pages).toBe(1);
  });

  test('Subscribe to totals', done => {
    expect(connection.list('category', 7).total.entities).toBe(null);
    expect(connection.list('category', 7).total.pages).toBe(null);
    expect(connection.list('category', 7).total.fetched.entities).toBe(null);
    expect(connection.list('category', 7).total.fetched.pages).toBe(null);
    expect(connection.list('category', 7).page(1).total).toBe(null);
    autorun(() => {
      if (
        connection.list('category', 7).total.entities === 10 &&
        connection.list('category', 7).total.pages === 2 &&
        connection.list('category', 7).page(1).total === 5 &&
        connection.list('category', 7).total.fetched.entities === 5 &&
        connection.list('category', 7).total.fetched.pages === 1
      )
        done();
    });
    connection.addPage({
      type: 'category',
      id: 7,
      page: 1,
      result: resultFromCategory7,
      entities: entitiesFromCategory,
      total: { entities: 10, pages: 2 },
    });
  });
});
