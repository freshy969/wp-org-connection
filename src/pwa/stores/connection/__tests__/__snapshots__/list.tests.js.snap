// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store › List Add list. Request and succeed 1`] = `
Object {
  "fetching": true,
  "pageMap": Object {
    "2": Object {
      "entities": Array [],
      "fetching": true,
      "ready": false,
    },
  },
  "ready": false,
  "total": Object {
    "entities": null,
    "pages": null,
  },
}
`;

exports[`Store › List Add list. Request and succeed 2`] = `
Object {
  "fetching": false,
  "pageMap": Object {
    "2": Object {
      "entities": Array [
        60,
      ],
      "fetching": false,
      "ready": true,
    },
  },
  "ready": true,
  "total": Object {
    "entities": 250,
    "pages": 25,
  },
}
`;

exports[`Store › List Check lists totals 1`] = `
Object {
  "fetching": false,
  "pageMap": Object {
    "0": Object {
      "entities": Array [
        60,
        61,
        62,
        63,
      ],
      "fetching": false,
      "ready": false,
    },
    "1": Object {
      "entities": Array [
        64,
        65,
        66,
      ],
      "fetching": false,
      "ready": false,
    },
  },
  "ready": false,
  "total": Object {
    "entities": 17,
    "pages": 5,
  },
}
`;

exports[`Store › List Retrieve list items 1`] = `
Object {
  "fetching": false,
  "pageMap": Object {
    "0": Object {
      "entities": Array [
        60,
        61,
      ],
      "fetching": false,
      "ready": false,
    },
    "1": Object {
      "entities": Array [
        62,
        63,
      ],
      "fetching": false,
      "ready": false,
    },
    "8": Object {
      "entities": Array [
        60,
        63,
      ],
      "fetching": false,
      "ready": false,
    },
  },
  "ready": false,
  "total": Object {
    "entities": null,
    "pages": null,
  },
}
`;
